sync_client.py

```diff
@@ -0,0 +1,231 @@
+# This is a proof of concept for demonstration only.  Co-written with ChatGPT.
+
+import argparse
+import gzip
+import hashlib
+import io
+import json
+import logging
+import os
+import time
+from collections import defaultdict
+
+import requests
+from sqlalchemy import Column, String, create_engine
+from sqlalchemy.dialects.sqlite import JSON as SQLiteJSON
+from sqlalchemy.orm import declarative_base, sessionmaker
+
+# Configure logging
+logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
+
+Base = declarative_base()
+
+
+def json_version(d: dict) -> str:
+    j = json.dumps(d, sort_keys=True)
+    s = j.encode("utf-8")
+    h = hashlib.sha256(s).hexdigest()
+    return h
+
+
+def gzipped_size(data: bytes) -> int:
+    buf = io.BytesIO()
+    with gzip.GzipFile(fileobj=buf, mode="wb") as f:
+        f.write(data)
+    return len(buf.getvalue())
+
+
+def log_json_preview(label: str, b: bytes):
+    try:
+        parsed = json.loads(b.decode("utf-8"))
+        pretty = json.dumps(parsed, indent=2)[:1024]
+        logging.debug(f"{label} preview (pretty-printed):\n{pretty}")
+    except Exception:
+        logging.debug(f"{label} preview (raw): {b[:1024]!r}")
+
+
+class Source(Base):
+    __tablename__ = "sources"
+    uuid = Column(String, primary_key=True)
+    data = Column(SQLiteJSON, nullable=True)
+
+
+class Item(Base):
+    __tablename__ = "items"
+    uuid = Column(String, primary_key=True)
+    data = Column(SQLiteJSON, nullable=True)
+
+
+# Set up SQLAlchemy engine and session
+engine = create_engine("sqlite:///sources.db")
+Base.metadata.create_all(engine)
+Session = sessionmaker(bind=engine)
+session = Session()
+
+# Command-line argument parsing
+parser = argparse.ArgumentParser()
+parser.add_argument("--prefix", type=int, help="Group source UUIDs by prefix of given length")
+parser.add_argument("-v", "--verbose", action="store_true", help="Enable debug logging")
+args = parser.parse_args()
+
+if args.verbose:
+    logging.getLogger().setLevel(logging.DEBUG)
+
+if args.prefix is not None:
+    grouped = defaultdict(int)
+    for source in session.query(Source):
+        prefix = source.uuid[: args.prefix]
+        grouped[prefix] += 1
+    for prefix, count in sorted(grouped.items()):
+        print(f"{prefix}: {count}")
+    session.close()
+    exit()
+
+api_host = os.getenv("SD_JOURNALIST_API", "localhost:8081")
+base_url = f"http://{api_host}/api/v1"
+
+# Construct local {uuid: version} for sources
+local_source_versions = {
+    # TODO: We should derive a fresh Source.collection_version, so that any
```

Comment:
+1, can definitely cache on write in the client too (similar to some of the server implementation in this now!)